src/Microsoft.AspNet.Mvc.Core/SessionStateTempDataProvider.cs

```diff
@@ -37,7 +65,27 @@ public class SessionStateTempDataProvider : ITempDataProvider
                 using (var memoryStream = new MemoryStream(value))
                 using (var writer = new BsonReader(memoryStream))
                 {
-                    tempDataDictionary = jsonSerializer.Deserialize<Dictionary<string, object>>(writer);
+                    tempDataDictionary = _jsonSerializer.Deserialize<Dictionary<string, object>>(writer);
+                }
+                foreach (var item in tempDataDictionary.ToList())
+                {
+                    var jArrayValue = item.Value as JArray;
+                    if (jArrayValue != null && jArrayValue.Count > 0)
+                    {
+                        Type returnType = null;
+                        _arrayTypeLookup.TryGetValue(jArrayValue[0].Type, out returnType);
+                        if (returnType != null)
```

Comment:
Can just do the `if()` on the return value of TryGetValue, e.g.: `if (_foo.TryGetValue(x, out y)) { ... do stuff with y ... }`.